src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -721,43 +729,50 @@ impl Tree {
     /// `child: Reserved`. This tree can exist. If we blindly delete `parent` and reassign
     /// `child` to be a direct child of `root` then Writes to `child` are now permitted
     /// whereas they were not when `parent` was still there.
-    fn keep_only_needed(&mut self, idx: UniIndex, live: &FxHashSet<BorTag>) -> bool {
-        let node = self.nodes.get(idx).unwrap();
-        // FIXME: this function does a lot of cloning, a 2-pass approach is possibly
-        // more efficient. It could consist of
-        // 1. traverse the Tree, collect all useless tags in a Vec
-        // 2. traverse the Vec, remove all tags previously selected
-        // Bench it.
-        let children: SmallVec<_> = node
-            .children
-            .clone()
-            .into_iter()
-            .filter(|child| self.keep_only_needed(*child, live))
-            .collect();
-        let no_children = children.is_empty();
-        let node = self.nodes.get_mut(idx).unwrap();
-        node.children = children;
-        if !live.contains(&node.tag) && no_children {
-            // All of the children and this node are unreachable, delete this tag
-            // from the tree (the children have already been deleted by recursive
-            // calls).
-            // Due to the API of UniMap we must absolutely call
-            // `UniValMap::remove` for the key of this tag on *all* maps that used it
-            // (which are `self.nodes` and every range of `self.rperms`)
-            // before we can safely apply `UniValMap::forget` to truly remove
-            // the tag from the mapping.
-            let tag = node.tag;
-            self.nodes.remove(idx);
-            for (_perms_range, perms) in self.rperms.iter_mut_all() {
-                perms.remove(idx);
+    fn keep_only_needed(&mut self, root: UniIndex, live: &FxHashSet<BorTag>) -> bool {
+        let mut stack = vec![(root, 0)];
+        // To avoid stack overflows, we roll our own stack.
+        // Each frame consists of the current tag, and the number of the next child.
+        while let Some((tag, nth_child)) = stack.last_mut() {
+            let node = self.nodes.get(*tag).unwrap();
+            if *nth_child == node.children.len() {
+                // We have visited all children. Let's remove those that are useless.
+                // Split the children into useful and useless.
+                let (useless, new_children) =
```

Comment:
```suggestion
                // FIXME: this copies new_children, even if nothing gets removed.
                let (useless, new_children) =
```

Reply:
This would presumably also need `Vec::drain_filter` which is currently unstable.

Reply:
Yeah, though we're using other unstable features in Miri as well so that wouldn't be a blocker.

Reply:
I've played around with the filter method. Sadly it requires two `mem::swap`s to make it work. See specifically [this commit](https://github.com/rust-lang/miri/pull/3833/commits/c9c21b5a7145616bdf980e6a7d7b72d56b46ec50).